Configurable resources for RGB and NN

GergelySzabolcs opened the pull request:
New options were added:
`sh/shaves` - number of shaves used by NN
`cmx/cmx_slices` -number of cmx slices used by NN
`NCE/NCEs` -number of NCEs used by NN
`rgb/rgb_cam_config` - rgb camera config, 1080p_30hz / 4k_30hz supported currently

Install requirements for cloud compile:
`python3 -m pip install -r requirements.txt`
Install dependencies for hostapp cpp backend:
```
Ubuntu:
        sudo apt-get install -y  libcurl4-openssl-dev
macOS:
        brew install curl
```

Run as :
./depthai.py -s metaout previewout -sh 8 -cmx 8 -NCE 1 -rgb 4k_30hz -v test.h264 -dd

Tested on Mac and Linux.

alex-luxonis commented on depthai.py:
```diff
@@ -333,6 +340,7 @@ def show_landmarks_recognition(entries_prev, frame):
         'store_to_eeprom': args['store_eeprom'],
         'clear_eeprom': args['clear_eeprom'],
         'override_eeprom': args['override_eeprom'],
+        'rgb_cam_config' : args['rgb_cam_config']
```
Was thinking of something like, related also to the other comment below:
```
    'camera':
    {
        'rgb':
        {
            # 3840x2160, 1920x1080
            'resolution_h': 1080,
            'fps': 30,
        },
        'mono':
        {
            # 1280x720, 640x400 (binning enabled)
            'resolution_h': 720,
            'fps': 45,
        },
    },
```
But may also need to add a category like `rgb_still`, for full 12MP capture.
Thoughts on a better approach?

GergelySzabolcs replied:
I tried to match the available sensor config modes from device. (mode 0, 1 etc.)
We already have `max_fps` option, should we remove that and add it here, or we keep both?

Luxonis-Brandon commented:
Thanks good idea.

GergelySzabolcs commented:
Added model downloader into cpp backend:
New dependencies to build hostapp:
```
Ubuntu:
        sudo apt-get install -y  libcurl4-openssl-dev
  - macOS:
        brew install curl
```

GergelySzabolcs commented:
Added new config structure suggested by @alex-luxonis 
Also the pipeline scheduler (between RGB and depth) is now dynamically configured.

Luxonis-Brandon commented:
Testing this I did run into the following at least once:
`Traceback (most recent call last):
  File "depthai.py", line 545, in <module>
    show_nn(nnet_prev["entries_prev"], frame, is_depth=True)
TypeError: show_mobilenet_ssd() missing 1 required positional argument: 'labels'`

When running `python3 test.py -s metaout previewout depth_sipp,5 -bb` on a BW1098OBC on my Mac.  

Full output:
```
python3 test.py -s metaout previewout depth_sipp,5 -bb
python3 depthai.py '-s' 'metaout' 'previewout' 'depth_sipp,5' '-bb' 
No calibration file. Using Calibration Defaults.
Using Arguments= {'config_overwrite': None, 'board': None, 'shaves': None, 'cmx_slices': None, 'NCEs': None, 'rgb_resolution': 1080, 'rgb_fps': 30, 'mono_resolution': 720, 'mono_fps': 30, 'field_of_view': 71.86, 'rgb_field_of_view': 68.7938, 'baseline': 9.0, 'rgb_baseline': 2.0, 'swap_lr': True, 'store_eeprom': False, 'clear_eeprom': False, 'override_eeprom': False, 'device_id': '', 'dev_debug': None, 'force_usb2': None, 'cnn_model': 'mobilenet-ssd', 'disable_depth': False, 'draw_bb_depth': True, 'full_fov_nn': False, 'streams': [{'name': 'metaout'}, {'name': 'previewout'}, {'name': 'depth_sipp', 'max_fps': 5.0}], 'video': None}
depthai.__version__ == 0.1.0
depthai.__dev_version__ == 2986ff5172fe8a9f95d375fa3fe2d6ff370863bb
XLink initialized.
Sending device firmware "cmd_file": /Users/leeroy/depthai/depthai.cmd
Successfully connected to device.
Loading config file
Attempting to open stream config_d2h
watchdog started 3000
Successfully opened stream config_d2h with ID #0!
Closing stream config_d2h: ...
Closing stream config_d2h: DONE.
WARNING: Version (dev) does not match (a0c257403d77f93573cd3c5d959f79b802541470 & 2986ff5172fe8a9f95d375fa3fe2d6ff370863bb)
EEPROM data: valid (v2)
  Board name     : BW1098OBC
  Board rev      : R0M0E0
  HFOV L/R       : 71.86 deg
  HFOV RGB       : 68.7938 deg
  L-R   distance : 7.5 cm
  L-RGB distance : 3.75 cm
  L/R swapped    : yes
  L/R crop region: top
  Calibration homography:
    1.002324,   -0.004016,   -0.552212,
    0.001249,    0.993829,   -1.710247,
    0.000008,   -0.000010,    1.000000,
Available streams: ['meta_d2h', 'left', 'right', 'disparity', 'depth_sipp', 'metaout', 'previewout', 'jpegout', 'video', 'object_tracker']
CNN configurations read: /Users/leeroy/depthai/resources/nn/mobilenet-ssd/mobilenet-ssd_depth.json
depthai: Calibration file is not specified, will use default setting;
config_h2d json:
{"_board":{"_homography_right_to_left":[0.988068163394928,0.0029474012553691864,5.067617416381836,-0.008765067905187607,0.9921473264694214,-8.795275688171387,-8.449587767245248e-06,-3.603489403758431e-06,1.0]},"_load_inBlob":true,"_pipeline":{"_streams":[{"name":"metaout"},{"name":"previewout"},{"data_type":"uint16","max_fps":5.0,"name":"depth_sipp"}]},"ai":{"NCEs":1,"blob_size":14514432,"calc_dist_to_bb":true,"cmx_slices":4,"keep_aspect_ratio":true,"shaves":4},"board":{"clear-eeprom":false,"left_fov_deg":71.86000061035156,"left_to_rgb_distance_m":0.019999999552965164,"left_to_right_distance_m":0.09000000357627869,"name":"","override-eeprom":false,"revision":"","rgb_fov_deg":68.7938003540039,"stereo_center_crop":false,"store-to-eeprom":false,"swap-left-and-right-cameras":true},"camera":{"mono":{"fps":30,"resolution_h":720},"rgb":{"fps":30,"resolution_h":1080}},"depth":{"confidence_threshold":0.5,"depth_limit_mm":10000,"padding_factor":0.30000001192092896},"ot":{"confidence_threshold":0.5,"max_tracklets":20}}
Attempting to open stream config_h2d
Successfully opened stream config_h2d with ID #1!
Writing 5000 bytes to config_h2d
!!! XLink write successful: config_h2d (5000)
Closing stream config_h2d: ...
Closing stream config_h2d: DONE.
Creating observer stream host_capture: ...
Attempting to open stream host_capture
Successfully opened stream host_capture with ID #0!
Creating observer stream host_capture: DONE.
Read: 14514432
Attempting to open stream inBlob
Successfully opened stream inBlob with ID #1!
Writing 14514432 bytes to inBlob
!!! XLink write successful: inBlob (14514432)
Closing stream inBlob: ...
Closing stream inBlob: DONE.
depthai: done sending Blob file /Users/leeroy/depthai/resources/nn/mobilenet-ssd/mobilenet-ssd.blob
Attempting to open stream outBlob
Successfully opened stream outBlob with ID #2!
Closing stream outBlob: ...
Closing stream outBlob: DONE.
CNN input width: 300
CNN input height: 300
CNN input num channels: 3
CNN to depth bounding-box mapping: start(318, 38), max_size(643, 643)
Host stream start:depth_sipp
Opening stream for read: depth_sipp
Attempting to open stream depth_sipp
Successfully opened stream depth_sipp with ID #1!
Starting thread for stream: depth_sipp
Started thread for stream: depth_sipp
Host stream start:metaout
Opening stream for read: metaout
Attempting to open stream metaout
Successfully opened stream metaout with ID #2!
Starting thread for stream: metaout
Started thread for stream: metaout
Host stream start:previewout
Opening stream for read: previewout
Attempting to open stream previewout
Successfully opened stream previewout with ID #3!
Starting thread for stream: previewout
Started thread for stream: previewout
depthai: INIT OK!
Traceback (most recent call last):
  File "depthai.py", line 545, in <module>
    show_nn(nnet_prev["entries_prev"], frame, is_depth=True)
TypeError: show_mobilenet_ssd() missing 1 required positional argument: 'labels'
Stopping threads: ...
Closing stream metaout: ...
Closing stream metaout: DONE.
Thread for metaout finished.
Closing stream previewout: ...
Closing stream previewout: DONE.
Thread for previewout finished.
Closing stream depth_sipp: ...
Closing stream depth_sipp: DONE.
Thread for depth_sipp finished.
Stopping threads: DONE 0.094s.
Closing all observer streams: ...
Closing stream host_capture: ...
Closing stream host_capture: DONE.
Closing all observer streams: DONE.
Reseting device: 0.
E: [xLink] [    627919] [] DispatcherWaitEventComplete:380	waiting is timeout, sending reset remote event
E: [xLink] [    627919] [] DispatcherAddEvent:323	curr == NULL expression failed
Return code:-11

```

Luxonis-Brandon commented:
Yes, I can confirm this happens every time @szabi-luxonis . ^

GergelySzabolcs commented:
Fixed @Luxonis-Brandon 
I was about to move NN decodings in different file, so it's cleaner, so I added a new parameter "labels", but I forget to update the code at one place.

Luxonis-Brandon commented:
Thanks @szabi-luxonis .  Yes, tested it and I can confirm it looks great!  No problem now on my Mac for the same command.
![image](https://user-images.githubusercontent.com/32992551/86054785-025f4180-ba18-11ea-911d-b4c78518e73f.png)

alex-luxonis commented on resources/nn/tiny-yolo/tiny-yolo.json:
```diff
@@ -0,0 +1,43 @@
+        [
+            "mask",
+            "no mask",
+            "no mask2"
```
Should we rename the model name `tiny-yolo` to include something about the face mask?

GergelySzabolcs replied:
good question. I think we should remove it, since it's a custom network, and maybe add the official one ?

GergelySzabolcs replied:
thoughts @Luxonis-Brandon  ?

Luxonis-Brandon replied:
Yes agreed.

alex-luxonis commented on model_compiler/model_converter.py:
```diff
@@ -0,0 +1,57 @@
+
+PLATFORM="VPU_MYRIAD_2450" if args['NCEs'] == 0 else "VPU_MYRIAD_2480"
+
+url = "http://69.164.214.171:8080/"
```
That's luxonis.com, right? Should we change the url to http://luxonis.com:8080 ?

alex-luxonis commented:
With a configuration that overloads the device so it can't keep up with 30fps processing, I see frame tearing or out-of-order frames, seems to appear in both previewout and h.264 streams, so could be from the ISP/post-processed output.
```
./depthai.py -sh 8 -cmx 8 -NCE 2 -rgbr 2160 -dd -v test.h264
./depthai.py -sh 8 -cmx 8 -NCE 2 -rgbr 2160 -dd
```
Not sure if caused by the change in frame update policy.
If not easy to resolve now, we could just flag the issue, as this may be an unusual configuration.

Luxonis-Brandon commented:
Yes agreed on if not easy to solve to just flag it as I agree it is likely an unusual configuration.

alex-luxonis commented on depthai_helpers/cli_utils.py:
```diff
@@ -44,6 +44,20 @@ def parse_args():
+                        help="Number of shaves used by NN.")
+    parser.add_argument("-cmx", "--cmx_slices", default=None, type=int,
+                        help="Number of cmx slices used by NN.")
+    parser.add_argument("-NCE", "--NCEs", default=None, type=int,
```
Minor thing, should we rename the short version of this argument `"-NCE"` to lowercase, to be consistent with the other arguments?
